src/circuit_proof/r1cs.rs

```diff
@@ -0,0 +1,649 @@
+        let var_b = cs.alloc_variable(Scalar::from_u64(4));
+        let var_c = cs.alloc_variable(Scalar::from_u64(13));
+        let var_d = cs.alloc_variable(Scalar::one());
+
+        let lc_a = LinearCombination::construct(
+            vec![
+                (var_a, Scalar::one()),
+                (var_b, Scalar::one()),
+                (var_c, -Scalar::one()),
+            ],
+            Scalar::from_u64(8),
+        );
+        let lc_b = LinearCombination::construct(vec![], Scalar::from_u64(2));
+        let lc_c = LinearCombination::construct(vec![(var_d, Scalar::from_u64(3))], Scalar::zero());
+        assert!(cs.push_lc(lc_a, lc_b, lc_c).is_ok());
+
+        let (circuit, prover_input, verifier_input) =
+            cs.create_proof_input(&pedersen_generators, &mut rng);
+        assert!(create_and_verify_helper(circuit, prover_input, verifier_input).is_err());
+    }
+
+    // Creates a 2 in 2 out shuffle circuit.
+    fn shuffle_circuit_helper(
+        in_0: Scalar,
+        in_1: Scalar,
+        out_0: Scalar,
+        out_1: Scalar,
+    ) -> Result<(), &'static str> {
+        let mut rng = OsRng::new().unwrap();
+        let pedersen_generators = PedersenGenerators::default();
+        let mut cs = ConstraintSystem::new();
+        let z = Scalar::random(&mut rng);
+
+        let var_in_0 = cs.alloc_variable(in_0);
+        let var_in_1 = cs.alloc_variable(in_1);
+        let var_out_0 = cs.alloc_variable(out_0);
+        let var_out_1 = cs.alloc_variable(out_1);
+        let var_mul = cs.alloc_variable((in_0 - z) * (in_1 - z));
+
+        // lc_0: (var_in_0 - z) * (var_in_1 - z) = var_mul
+        let lc_0_a = LinearCombination::construct(vec![(var_in_0, Scalar::one())], -z);
+        let lc_0_b = LinearCombination::construct(vec![(var_in_1, Scalar::one())], -z);
+        let lc_0_c =
+            LinearCombination::construct(vec![(var_mul.clone(), Scalar::one())], Scalar::zero());
+        assert!(cs.push_lc(lc_0_a, lc_0_b, lc_0_c).is_ok());
+
+        // lc_1: (var_out_0 - z) * (var_out_1 - z) = var_mul
+        let lc_1_a = LinearCombination::construct(vec![(var_out_0, Scalar::one())], -z);
+        let lc_1_b = LinearCombination::construct(vec![(var_out_1, Scalar::one())], -z);
+        let lc_1_c = LinearCombination::construct(vec![(var_mul, Scalar::one())], Scalar::zero());
+        assert!(cs.push_lc(lc_1_a, lc_1_b, lc_1_c).is_ok());
+
+        let (circuit, prover_input, verifier_input) =
+            cs.create_proof_input(&pedersen_generators, &mut rng);
+        create_and_verify_helper(circuit, prover_input, verifier_input)
+    }
+
+    #[test]
+    // Test that a 2 in 2 out shuffle circuit behaves as expected
+    fn shuffle_circuit() {
+        let three = Scalar::from_u64(3);
+        let seven = Scalar::from_u64(7);
+        assert!(shuffle_circuit_helper(three, seven, seven, three).is_ok());
+        assert!(shuffle_circuit_helper(three, seven, seven, three).is_ok());
```

Comment:
These two lines look identical. Did you mean to permute the order of the last two arguments?

Reply:
whoops - yes, I'll fix that. good catch!